### ISSUE TITLE
[Discussion] : delegate to other package managers

### ISSUE THREAD

**Issue by alalazo** [LLNL/spack#1249]
Many extensions for a given language follow a well defined structure. For instance : 
```python
class RMunsell(Package):
    homepage = "https://cran.r-project.org/web/packages/munsell/index.html"
    url      = "https://cran.r-project.org/src/contrib/munsell_0.4.3.tar.gz"
    list_url = "https://cran.r-project.org/src/contrib/Archive/munsell"

    version('0.4.3', 'ebd205323dc37c948f499ee08be9c476')

    extends('R')

    depends_on('r-colorspace')

    def install(self, spec, prefix):
        R('CMD', 'INSTALL', '--library={0}'.format(self.module.r_lib_dir),
          self.stage.source_path)
```
Changing class name,  url, version and list of dependencies we may obtain another package :
```python
class RMime(Package):
    homepage = "https://github.com/yihui/mime"
    url      = "https://cran.r-project.org/src/contrib/mime_0.4.tar.gz"
    list_url = "https://cran.r-project.org/src/contrib/Archive/mime"

    version('0.4', '789cb33e41db2206c6fc7c3e9fbc2c02')

    extends('R')

    def install(self, spec, prefix):
        R('CMD', 'INSTALL', '--library={0}'.format(self.module.r_lib_dir),
          self.stage.source_path)
```

So I was wondering : why not having a package factory used at install time that generates / updates those kind of things for us ? 

The (still vague) idea I have in mind is basically : 

 1. if a package exists explicitly in any registered repo (like `builtin`), use it
 2. if not and it matches an existing naming scheme (for instance `r-*` or `py-*`)
   * try to generate automatically the corresponding `package.py`
   * save it in a separate repository for autogenerated packages.py (`py-builtin` or something similar)
   * register the repository (if not there before) giving it the least precedence

This proposal requires of course further analysis, and it certainly demands the extendee package manager to be queriable, but I was wondering at a high level what other people thinks about it.

If I didn't miss obvious points it seems to me that an approach like this would give two major benefits : 
- once in place for a language, it will give automatic access to most of its packages
- it will reduce the lines of code to be maintained in `var\spack\repos\builtin` by a fair amount (as only the packages that really need a customization would reside there, e.g. `py-numpy` and similar)

**Comment by citibeth** [comment]
This is simlar to #1151 --- the idea that Spack would auto-generate packages in some cases, and be clever enough doing so that we can trust them the first time without hand-checking them.

> it will reduce the lines of code to be maintained in var\spack\repos\builtin by a fair amount (as only the packages that really need a customization would reside there, e.g. py-numpy and similar)

There needs to be a way to indicate to Spack that "somebody has checked that this package is correctly auto-generated," not just "nobody has looked at this package yet."

**Comment by adamjstewart** [comment]
`spack create` already does this. It doesn't setup the homepage or list_url for you, but I could easily add those for specific build systems if you want.

**Comment by alalazo** [comment]
@adamjstewart I think that `create` (or the code in it) will be reused in large part if we pursue something like this. What is currently missing is mainly the logic to query another packaging system, for instance to write the correct list of dependencies and similar things.

**Comment by adamjstewart** [comment]
Oh, adding dependencies would be an interesting idea.

**Comment by tgamblin** [comment]
I don't think this is a bad idea, I just wonder how much work it is to do it well.  Getting everything right takes some effort for a most packages (although there are a lot of really simple ones).  Also there are things like variants and optional dependencies that other package managers aren't always going to expose to you in a convenient way.  Seems worth looking at but might be hard to get right and maintain.

**Comment by trws** [comment]
Agree with @tgamblin for the most part, though some would be easier than others.  LuaRocks and go get packages, for example, would be pretty easy to do this way since they all (by necessity) work the same constrained way.  Managers like pip, npm, etc. I would expect to be rather harder though.

**Comment by alalazo** [comment]
@TomasPuverle I saw your message "Spack and (Ana)Conda - living side by side" in the mailing list and thought you might be interested in this issue and the ones referenced above

**Comment by TomasPuverle** [comment]
Thank you for pulling me into the discussion; this is exactly what I was referring to in the thread on google.

I am having a bit of trouble understanding the consensus here; the way I'm reading the issue, I am seeing "this seems worthwhile but hard".

I have a question, especially wrt dependencies:  Why is it important for spack to have the "foreign" dependencies?  I understand that in an ideal scenario, this may be useful metadata, but from a practical perspective, it wouldn't really matter as long as the other manager does its job, no?  Sorry if there's something I'm not immediately grasping here.

(As an aside, let me paraphrase something I wrote earlier on the google forums) 
Spack sort of already has a hard-coded concept of this for "locally installed packages" using the packages.yaml concept.  In the proposed "abstract repo" case, the local packages (effectively the @system stuff) would just become another repo.  This seems like a clean abstraction to me.  Note that you also don't have dependency information in that case, as the packages were installed and handled by and "external" package manager - yum/dnf/apt or something similar.  I think there may be value in treating e.g. pip, conda and others in the same, blackbox-ish, way.

**Comment by citibeth** [comment]
There is a difference between Spack delegating to other package managers (getting them to build its packages, or analyzing their package repos to auto-generate Spack package); and linking a Spack repo with another repo.

**Comment by citibeth** [comment]
> Spack sort of already has a hard-coded concept of this for "locally installed packages" using the packages.yaml concept

I do like the idea of a more formalize Spack repo that consists of foreign-installed packages; rather than mixing this into `packages.yaml` the way it's done now.

**Comment by alalazo** [comment]
For future reference: pypa API is described in [PEP-503](https://www.python.org/dev/peps/pep-0503/)

**Comment by alalazo** [comment]
Other discussion on DOAP (Description Of A Project): #2282

**Comment by baberlevi** [comment]
R packages could be templated from the DESCRIPTION file:
https://cran.r-project.org/doc/manuals/r-release/R-exts.html#The-DESCRIPTION-file

**Comment by hartzell** [comment]
[I came here from #10837.]

A partial step towards this might be a tool for package authors.

E.g. [Homebrew's "pypi-poet"][pypi-poet], which

>Generates Homebrew resource stanzas for Python packages.

It could be interesting/help to do something similar for `go` and perhaps `rust`.

[pypi-poet]: https://github.com/tdsmith/homebrew-pypi-poet

**Comment by AndrewGaspar** [comment]
FYI #16185 is tackling some of these issues, specifically related to Cargo. It's not attempting to automatically generate the entirety of the package, but it is able to pull directly from the most popular package repository ("crates.io"), generate a pretty complete package using "spack create", and automatically cache/vendor dependencies of the package via "crates.io".

Might be worth taking a look at it and let me know what's missing if you're interested in this.
